android/runtime/common/src/java/org/appcelerator/kroll/util/KrollAssetHelper.java

```diff
@@ -138,4 +250,117 @@ public static String getCacheDir()
 	{
 		return cacheDir;
 	}
+
+	private static String normalizeDirectoryPath(String path)
+	{
+		// Normalize the given path, if necessary.
+		if ((path == null) || path.equals("/")) {
+			// Path references the root directory.
+			path = "";
+		} else if (!path.isEmpty() && !path.endsWith("/")) {
+			// Subdirectory paths must end with a slash.
+			path += "/";
+		}
+		return path;
+	}
+
+	private static String normalizeAssetCryptPath(String path)
+	{
+		final String ROOT_PATH_NAME = "Resources/";
+
+		if (path != null) {
+			if (path.startsWith(ROOT_PATH_NAME)) {
+				path = path.substring(ROOT_PATH_NAME.length());
+			}
+		}
+		return path;
+	}
+
+	/**
+	 * Stores a mapping of directory paths and their file/subfolder listings.
+	 * <p>
+	 * The keys are the full relative path to the directory. These directory paths always end with
+	 * a trailing slash except for the root directory which is represented by an empty string.
+	 * <p>
+	 * The values store a listing of files and subdirectories owned by the key's directory path.
+	 * <p>
+	 * You are expected to call this class' addPath() method to extract directories from the given
+	 * path and add its trailing file to the listing.
+	 */
+	private static class DirectoryListingMap extends HashMap<String, TreeSet<String>>
+	{
+		public DirectoryListingMap(int initialCapacity)
+		{
+			super(initialCapacity);
+		}
+
+		public void addPath(String path)
+		{
+			// Null path references the root directory.
+			if (path == null) {
+				path = "";
+			}
+
+			// Remove path separator(s) from the front of the path. It must be relative down below.
+			int startIndex = 0;
+			for (; (startIndex < path.length()) && (path.charAt(startIndex) == '/'); startIndex++);
```

Comment:
Couldn't we just do:
```Java
int startIndex = path.indexOf('/');
```
?

Reply:
I'm doing this in case there are 2 (or more) slashes "//" next to each other. Kind of like what Hans is asking me to do above with `$PROJECT_DIR`. :)